nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core/src/main/java/org/apache/nifi/controller/leader/election/CuratorLeaderElectionManager.java

```diff
@@ -64,10 +64,10 @@
 
     private volatile boolean stopped = true;
 
-    private final Map<String, LeaderRole> leaderRoles = new HashMap<>();
-    private final Map<String, RegisteredRole> registeredRoles = new HashMap<>();
+    private final ConcurrentMap<String, LeaderRole> leaderRoles = new ConcurrentHashMap<>();
+    private final ConcurrentMap<String, RegisteredRole> registeredRoles = new ConcurrentHashMap<>();
 
-    private final Map<String, TimedBuffer<TimestampedLong>> leaderChanges = new HashMap<>();
+    private final ConcurrentMap<String, TimedBuffer<TimestampedLong>> leaderChanges = new ConcurrentHashMap<>();
```

Comment:
Access to these Maps does not appear to have changed, so it is necessary to change the property type from `Map` to `ConcurrentMap`, or would it be sufficient to just change the implementation from `HashMap` to `ConcurrentHashMap`?

Reply:
@exceptionfactory I was considering commenting the same thing. But then I decided not to because I think there's some value in calling it a `ConcurrentMap`. Similarly, if you instantiated `TreeSet` but didn't use any of the methods in `SortedSet`, only those in `Set` would you declare it as `Set` or `SortedSet`? I think there's some value in declaring that a `SortedSet` just because the declaration makes it clearer of your expectations. Similarly here, it makes it more clear that the declared `Map` is expected to be thread safe.

That said, I don't have a problem with it one way or the other but would tend to lean slightly toward `ConcurrentMap`.

Reply:
Thanks for the reply @markap14, that's fair, especially given the use case and associated issue. I'm fine with leaving it as it stands, but appreciate the consideration.